src/OpenTelemetry.Exporter.Prometheus/Implementation/PrometheusMetricBuilder.cs

```diff
@@ -0,0 +1,276 @@
+                    writer.Write(@"}");
+                }
+
+                // value is a float represented as required by Go's ParseFloat() function. In addition to
+                // standard numerical values, Nan, +Inf, and -Inf are valid values representing not a number,
+                // positive infinity, and negative infinity, respectively.
+                writer.Write(" ");
+                writer.Write(m.Value.ToString(CultureInfo.InvariantCulture));
+                writer.Write(" ");
+
+                // The timestamp is an int64 (milliseconds since epoch, i.e. 1970-01-01 00:00:00 UTC, excluding
+                // leap seconds), represented as required by Go's ParseInt() function.
+                writer.Write(now);
+
+                // Prometheus' text-based format is line oriented. Lines are separated
+                // by a line feed character (\n). The last line must end with a line
+                // feed character. Empty lines are ignored.
+                writer.Write("\n");
+            }
+
+            static string GetLabelAndValue(string label, string value)
+            {
+                var safeKey = GetSafeLabelName(label);
+                var safeValue = GetSafeLabelValue(value);
+                return $"{safeKey}=\"{safeValue}\"";
+            }
+        }
+
+        private static string GetSafeName(string name, char[] firstCharNameCharset, char[] charNameCharset)
+        {
+            // https://prometheus.io/docs/concepts/data_model/#metric-names-and-labels
+            //
+            // Metric names and labels
+            // Every time series is uniquely identified by its metric name and a set of key-value pairs, also known as labels.
+            // The metric name specifies the general feature of a system that is measured (e.g. http_requests_total - the total number of HTTP requests received). It may contain ASCII letters and digits, as well as underscores and colons. It must match the regex [a-zA-Z_:][a-zA-Z0-9_:]*.
+            // Note: The colons are reserved for user defined recording rules. They should not be used by exporters or direct instrumentation.
+            // Labels enable Prometheus's dimensional data model: any given combination of labels for the same metric name identifies a particular dimensional instantiation of that metric (for example: all HTTP requests that used the method POST to the /api/tracks handler). The query language allows filtering and aggregation based on these dimensions. Changing any label value, including adding or removing a label, will create a new time series.
+            // Label names may contain ASCII letters, numbers, as well as underscores. They must match the regex [a-zA-Z_][a-zA-Z0-9_]*. Label names beginning with __ are reserved for internal use.
+            // Label values may contain any Unicode characters.
+
+            var sb = new StringBuilder();
```

Comment:
1. creating a string builder for each call could be very expensive, at minimum we need to cache it.
2. instead of doing the conversion each time, it might make sense to cache some frequently used value (e.g. consider some data structure such as splay tree).